packages/react-reconciler/src/ReactFiberWorkLoop.new.js

```diff
@@ -2361,6 +2366,14 @@ function commitLayoutEffects(
 ) {
   let fiber = firstChild;
   while (fiber !== null) {
+    let prevProfilerOnStack = null;
+    if (enableProfilerTimer && enableProfilerCommitHooks) {
+      if (fiber.tag === Profiler) {
```

Comment:
Can you lift the switch statement out of `commitLayoutEffects` so that we don't check this twice?

Reply:
I don't understand what you mean.

Are you suggesting I duplicate the code in between these two tag-type checks?

Reply:
I meant `commitLayoutEffectOnFiber`. It contains a `switch` on `fiber.tag`. We should combine that function with the outer, recursive function so that there's only a single check.

Reply:
I thought an explicit goal was to keep as much tag-specific logic inside of the begin/complete/commit/unwind modules.

I only lifted the duration bubbling out in this case because I had to.

Reply:
Yeah I would move the whole thing into `ReactFiberCommitWork`. The boundary between the modules is a bit awkward either way, but we can figure that out later. The more important thing for this PR is the boundary between the functions, IMO.

Reply:
I see. That still seems a little awkward, since I thought another explicit goal was to keep tree-traversal in the work loop.

Reply:
This suggestion still doesn't really make sense to me.

There are 3 functions (well, 3 for each– layout and passive):
* The outer loop function (e.g. `commitLayoutEffects`) where we maintain our stack
* The "impl" function (e.g. `commitLayoutEffectsImpl`)
* The inner function with the switch case (e.g. `commitLifeCycles`)

You're suggesting I lift the switch case from `commitLifeCycles` to the outer traversal function `commitLayoutEffects`, but the separate "impl" function exists so that we can use `invokeGuardedCallback` for error handling in DEV.

What am I overlooking?

Reply:
Let's set up some time to discuss over chat 👍

Reply:
> Can you lift the switch statement out of `commitLayoutEffects` so that we don't check this twice?

We're still going to end up checking `tag` twice, because of how we need to check during recursion (to manage the stack) and again when committing a specific fiber's effects. The only way around this would be move all of the the `tag` checks into the outer recursive function, but that would result in *a lot of duplication* since we'd end up needing to copy the recursive block into every case.
```js
// This would have to be copied into every switch/case
let child = finishedWork.child;
while (child !== null) {
  const primaryFlags = child.flags & LayoutMask;
  const primarySubtreeFlags = child.subtreeFlags & LayoutMask;
  if (primaryFlags !== NoFlags || primarySubtreeFlags !== NoFlags) {
    recursivelyCommitLayoutEffects(child, root, committedLanes);
  }
  child = child.sibling;
}
```

I think we're on the same page here, since even your [original pseudo-code example](https://github.com/facebook/react/pull/19862#discussion_r491161670) showed two sets of switch statements. The main difference between the pseudo-code and what I initially wrote is that I left the second switch/case inside of the `commitLayoutEffectsForFiber` function (rather than lifting it out into the recursive loop). My reasoning for doing this was to avoid having to have DEV checks with `invokeGuardedCallback` inside of every case.

That being said, I also committed a follow up which *did* lift the case up to the recursive function and just repeated the try/catch (or `invokeGuardedCallback`). I'm not sure which is better at this point. I think I'm too close to it. I'm going to commit and push both approaches separately so you can look and give a second opinion.

Note that both approaches currently have cause a failure in `ReactSuspenseFuzz-test` that I haven't yet tracked down:
> Warning: Can't perform a React state update on an unmounted component.

Also– do you think we should do a similar thing for the passive effect flushing methods? Your comment was about layout effects. Changing the passive effect methods would be a little bit of a larger change I think.

Reply:
> Note that both approaches currently have cause a failure in ReactSuspenseFuzz-test that I haven't yet tracked down:

I'm going to step away from this and go running. It's probably something silly but I'm not seeing it.